tools/syz-crush/crush.go

```diff
@@ -24,107 +28,252 @@ import (
+type CrashReport struct {
+	vmIndex int
+	Report  *report.Report
+}
+
+type FileType int
+
+const (
+	LogFile FileType = iota
+	CProg
+)
+
+func getType(fileName string) FileType {
+	extension := filepath.Ext(fileName)
+
+	switch extension {
+	case ".c":
+		return CProg
+	case ".txt", ".log":
+		return LogFile
+	default:
+		log.Logf(0, "assuming logfile type")
+		return LogFile
+	}
+}
+
 func main() {
 	flag.Parse()
 	cfg, err := mgrconfig.LoadFile(*flagConfig)
 	if err != nil {
 		log.Fatalf("%v", err)
+		os.Exit(1)
 	}
 	if len(flag.Args()) != 1 {
-		log.Fatalf("usage: syz-crush -config=config.file execution.log")
+		log.Fatalf("usage: syz-crush -config=config.file <execution.log|creprog.c>")
+	}
+
+	if err := osutil.MkdirAll(cfg.Workdir); err != nil {
+		log.Fatalf("failed to create tmp dir: %v", err)
+	}
+
+	if *flagInfinite {
+		log.Logf(0, "running infinitely and restarting VM every %v", *flagRestartTime)
+	} else {
+		log.Logf(0, "running until crash is found or till %v", *flagRestartTime)
 	}
-	if _, err := prog.GetTarget(cfg.TargetOS, cfg.TargetArch); err != nil {
+
+	target, err := prog.GetTarget(cfg.TargetOS, cfg.TargetArch)
+	if err != nil {
 		log.Fatalf("%v", err)
 	}
+
 	vmPool, err := vm.Create(cfg, false)
 	if err != nil {
 		log.Fatalf("%v", err)
 	}
+
 	reporter, err := report.NewReporter(cfg)
 	if err != nil {
 		log.Fatalf("%v", err)
 	}
 
-	log.Logf(0, "booting test machines...")
+	reproduceMe := flag.Args()[0]
+	baseName := filepath.Base(reproduceMe)
+
+	runType := getType(reproduceMe)
+	if runType == CProg {
+		execprog, err := ioutil.ReadFile(reproduceMe)
+		if err != nil {
+			log.Fatalf("error reading source file from '%s'", reproduceMe)
+		}
+
+		cfg.SyzExecprogBin, err = csource.BuildNoWarn(target, execprog)
+		if err != nil {
+			log.Fatalf("failed to build source file: %v", err)
+		}
+
+		log.Logf(0, "compiled csource %v to cprog: %v", reproduceMe, cfg.SyzExecprogBin)
+	} else {
+		log.Logf(0, "reproducing from logfile: %v", reproduceMe)
+	}
+
+	restartPeriod := *flagRestartTime
+	if restartPeriod == 0 {
+		// Set default restart period to 1h
+		restartPeriod = time.Hour
+	}
+	log.Logf(0, "restartTime set to: %v", *flagRestartTime)
+
+	log.Logf(0, "booting test machines... %v", vmPool.Count())
+	runDone := make(chan *CrashReport, vmPool.Count())
 	var shutdown uint32
-	var wg sync.WaitGroup
-	wg.Add(vmPool.Count() + 1)
+	var runningWorkers uint32
+
 	for i := 0; i < vmPool.Count(); i++ {
-		i := i
-		go func() {
-			defer wg.Done()
+		atomic.AddUint32(&runningWorkers, 1)
+		go func(index int) {
 			for {
-				runInstance(cfg, reporter, vmPool, i)
-				if atomic.LoadUint32(&shutdown) != 0 {
+				runDone <- runInstance(target, cfg, reporter, vmPool, index, *flagRestartTime,
+					runType)
+				if atomic.LoadUint32(&shutdown) != 0 || !*flagInfinite {
+					atomic.AddUint32(&runningWorkers, ^uint32(0))
+
+					// If this is the last worker then we can close the channel
+					if atomic.LoadUint32(&runningWorkers) == 0 {
```

Comment:
There is a race here.
Consider, 2 goroutines decrement runningWorkers to 0. Then both load and _both_ get 0, so both close the channel -> bang!